[WTF] Fix performance when creating Vectors from different constness

calvaris opened the pull request:
#### fb30ea0502c628aa0c469505cc754deac4b767c6
<pre>
[WTF] Fix performance when creating Vectors from different constness
<a href="https://bugs.webkit.org/show_bug.cgi?id=299008">https://bugs.webkit.org/show_bug.cgi?id=299008</a>

Reviewed by Chris Dumez.

For example, when you try to create a Vector&lt;uint8_t&gt; from a span&lt;const uint8_t&gt;, templates will consider both types
different and even when you pass true to the template, the second VectorCopier::uninitializedCopy function will be used
and will defer to the slow version.

My use case was in MSE, creating a SharedBuffer implies creating a FragmentedSharedBuffer and then a DataSegment that
implicitly creates a Vector&lt;uint8_t&gt; from a span&lt;const uint8_t&gt;. If you&apos;re copying 2.5MB, time is reduced to 33% of the
initial one.

* Source/WTF/wtf/Vector.h:

Canonical link: <a href="https://commits.webkit.org/300464@main">https://commits.webkit.org/300464@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d78c6ea22026ee523e2f3aec1284766bd9518b9f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/122624 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/42332 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/33017 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/129235 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/74728 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/43050 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/50925 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/93207 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/61897 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/125576 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/34328 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/109782 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/73853 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/33312 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/27940 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/72720 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/114784 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/104031 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/28149 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/131959 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/121150 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/49565 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/37720 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/101752 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/49940 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/106002 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/101615 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25800 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/46972 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/25134 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/46339 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/49423 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/55172 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/151309 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/48893 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/38712 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/52245 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/50572 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

calvaris commented:
According to @fujii this could be a regression on https://commits.webkit.org/276220@main .

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/b84b5bce50ef5e43192827835d29119c5dfa2888. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50849

weinig commented:
If you do move forward with this, please at an API test this.

cdumez reviewed (approved):
LGTM

calvaris committed:
[WTF] Fix performance when creating Vectors from different constness
https://bugs.webkit.org/show_bug.cgi?id=299008

Reviewed by Chris Dumez.

For example, when you try to create a Vector<uint8_t> from a span<const uint8_t>, templates will consider both types
different and even when you pass true to the template, the second VectorCopier::uninitializedCopy function will be used
and will defer to the slow version.

My use case was in MSE, creating a SharedBuffer implies creating a FragmentedSharedBuffer and then a DataSegment that
implicitly creates a Vector<uint8_t> from a span<const uint8_t>. If you're copying 2.5MB, time is reduced to 33% of the
initial one.

* Source/WTF/wtf/Vector.h:

Canonical link: https://commits.webkit.org/300464@main

aperezdc commented:
Backported into `webkitglib/2.50` as commit e6d818a4732a7cc80cc48d72ef50f5db6529d07b

aperezdc commented:
Backported into `webkitglib/2.46` as commit 857965a502463c315b09c4bd860c831f59e93ad4
